// @abgov/ui-components-common/index.js
var Z = Object.defineProperty;
var q = (n) => {
  throw TypeError(n);
};
var z = (n, t, e) => t in n ? Z(n, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : n[t] = e;
var w = (n, t, e) => z(n, typeof t != "symbol" ? t + "" : t, e), K = (n, t, e) => t.has(n) || q("Cannot " + e);
var R = (n, t, e) => t.has(n) ? q("Cannot add the same private member more than once") : t instanceof WeakSet ? t.add(n) : t.set(n, e);
var V = (n, t, e) => (K(n, t, "access private method"), e);
const Oe = {
  NO_SCROLL: "no-scroll",
  AT_TOP: "at-top",
  MIDDLE: "middle",
  AT_BOTTOM: "at-bottom"
};
class Se {
  constructor(t) {
    w(this, "validators");
    this.validators = t || {};
  }
  add(t, ...e) {
    this.validators[t] = e;
  }
  validate(t) {
    const e = {};
    return Object.entries(this.validators).forEach(([r, a]) => {
      const i = a.map((s) => s(t[r])).find((s) => !!s);
      i && (e[r] = i);
    }), e;
  }
}
function Pe() {
  return [
    F("Day is required"),
    A({
      min: 1,
      max: 31,
      minMsg: "Day must be between 1 and 31",
      maxMsg: "Day must be between 1 and 31"
    })
  ];
}
function ve() {
  return [
    F("Month is required"),
    A({
      min: 0,
      max: 11,
      minMsg: "Month must be between Jan and Dec",
      maxMsg: "Month must be between Jan and Dec"
    })
  ];
}
function xe() {
  const n = (/* @__PURE__ */ new Date()).getFullYear();
  return [
    F("Year is required"),
    A({
      min: 1900,
      max: n,
      minMsg: "Year must be greater than 1900",
      maxMsg: `Year must be less than ${n}`
    })
  ];
}
function F(n) {
  return (t) => (n = n || "Required", typeof t == "number" && !isNaN(t) || t ? "" : n);
}
function ke(n) {
  const t = new RegExp(/^\+?[\d-() ]{10,18}$/);
  return C(t, n || "Invalid phone number");
}
function Ne(n) {
  const t = new RegExp(
    /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
  );
  return C(t, n || "Invalid email address");
}
function Te() {
  return (n) => {
    if (!n) return "";
    const t = "121212121".split("").map((a) => parseInt(a)), e = n.replace(/\D/g, "");
    return e.length !== 9 ? "SIN must contain 9 numbers" : e.split("").map((a) => parseInt(a)).map((a, i) => {
      const s = a * t[i];
      return s < 10 ? s : `${s}`.split("").map((o) => parseInt(o)).reduce((o, c) => o + c, 0);
    }).reduce((a, i) => a + i, 0) % 10 === 0 ? "" : "Invalid SIN";
  };
}
function Ye() {
  return C(
    /^[ABCEGHJ-NPRSTVXY]\d[ABCEGHJ-NPRSTV-Z][ -]?\d[ABCEGHJ-NPRSTV-Z]\d$/i,
    "Invalid postal code"
  );
}
function C(n, t) {
  return (e) => !e || e.match(n) ? "" : t;
}
function We({
  invalidMsg: n,
  minMsg: t,
  maxMsg: e,
  min: r,
  max: a
} = {}) {
  return (i) => {
    let s = /* @__PURE__ */ new Date(-1);
    return `${i || ""}`.length === 0 ? "" : (typeof i == "string" && (s = new Date(i)), typeof i == "number" && (s = new Date(i)), i.toDateString && (s = i), s.getDate() === -1 ? n || "Invalid date" : r && s < r ? t || `Must be after ${r}` : a && s > a ? e || `Must be before ${a}` : "");
  };
}
function A({
  invalidTypeMsg: n,
  minMsg: t,
  maxMsg: e,
  min: r = -Number.MAX_VALUE,
  max: a = Number.MAX_VALUE
} = {}) {
  return (i) => {
    let s = Number.MAX_VALUE;
    return `${i ?? ""}`.length === 0 ? "" : (typeof i == "string" && (s = parseFloat(i)), typeof i == "number" && (s = i), isNaN(s) ? n || "Must be a numeric value" : s > a ? e || `Must be less than or equal to ${a}` : s < r ? t || `Must be greater than or equal to ${r}` : "");
  };
}
function _e({
  invalidTypeMsg: n,
  minMsg: t,
  maxMsg: e,
  min: r = -Number.MAX_VALUE,
  max: a = Number.MAX_VALUE
}) {
  return (i) => `${i || ""}`.length === 0 ? "" : typeof i != "string" ? n || "Invalid type" : i.length > a ? e || `Must be less than ${a} characters` : i.length < r ? t || `Must be greater than ${r} characters` : "";
}
function L(n, t, e, r) {
  if (!n) {
    console.error("dispatch element is null");
    return;
  }
  n.dispatchEvent(
    new CustomEvent(t, {
      composed: !0,
      bubbles: r == null ? void 0 : r.bubbles,
      detail: e
    })
  );
}
function M(n, t, e, r) {
  if (!n) {
    console.error("dispatch element is null");
    return;
  }
  n.dispatchEvent(
    new CustomEvent("msg", {
      composed: !0,
      bubbles: r == null ? void 0 : r.bubbles,
      detail: {
        action: t,
        data: e
      }
    })
  );
}
var N, tt, G;
class Ee {
  constructor(t) {
    R(this, N);
    w(this, "state");
    w(this, "_formData");
    w(this, "_formRef");
    w(this, "_isCompleting", !1);
    this.type = t;
  }
  // Obtain reference to the form element
  init(t) {
    if (this._formRef) {
      console.warn("init: form element has already been set");
      return;
    }
    this._formRef = t.detail.el, this.state = {
      uuid: crypto.randomUUID(),
      form: {},
      history: [],
      editting: "",
      status: "not-started"
    };
  }
  initList(t) {
    this._formRef = t.detail.el, this.state = [];
  }
  // Public method to allow for the initialization of the state
  initState(t, e) {
    M(this._formRef, "external::init:state", t), typeof t == "string" ? this.state = JSON.parse(t) : Array.isArray(t) || (this.state = t), e && setTimeout(e, 201);
  }
  updateListState(t) {
    const e = t.detail;
    Array.isArray(e.data) && (this.state = e.data);
  }
  updateObjectState(t) {
    var r, a;
    if (Array.isArray(this.state))
      return;
    const e = t.detail;
    e.type === "list" ? this.state = {
      ...this.state,
      form: { ...((r = this.state) == null ? void 0 : r.form) || {}, [e.id]: e.data }
    } : this.state = {
      ...this.state,
      ...e.data,
      form: { ...((a = this.state) == null ? void 0 : a.form) || {}, ...e.data.form },
      history: e.data.history
    };
  }
  getStateList() {
    return this.state ? Array.isArray(this.state) ? this.state.length === 0 ? [] : this.state.map((t) => Object.values(t.form).filter((e) => {
      var r;
      return ((r = e == null ? void 0 : e.data) == null ? void 0 : r.type) === "details";
    }).map((e) => {
      var r;
      return e.data.type === "details" && ((r = e.data) == null ? void 0 : r.fieldsets) || {};
    }).reduce(
      (e, r) => {
        for (const [a, i] of Object.entries(r))
          e[a] = i.value;
        return e;
      },
      {}
    )) : (console.warn(
      "Utils:getStateList: unable to update the state of a non-multi form type",
      this.state
    ), []) : [];
  }
  // getStateItems(group: string): Record<string, FieldsetItemState>[] {
  //   if (Array.isArray(this.state)) {
  //     console.error(
  //       "Utils:getStateItems: unable to update the state of a multi form type",
  //     );
  //     return [];
  //   }
  //   if (!this.state) {
  //     console.error("Utils:getStateItems: state has not yet been set");
  //     return [];
  //   }
  //
  //   const data = this.state.form[group].data;
  //   if (data.type !== "list") {
  //     return [];
  //   }
  //
  //   return data.items.;
  // }
  // Public method to allow for the retrieval of the state value
  getStateValue(t, e) {
    if (Array.isArray(this.state))
      return console.error("getStateValue: unable to update the state of a multi form type"), "";
    if (!this.state)
      return console.error("getStateValue: state has not yet been set"), "";
    const r = this.state.form[t].data;
    return r.type !== "details" ? "" : r.fieldsets[e].value;
  }
  // Public method to allow for the continuing to the next page
  continueTo(t) {
    if (!t) {
      console.error("continueTo [name] is undefined");
      return;
    }
    M(this._formRef, "external::continue", { next: t });
  }
  // Public method to perform validation and send the appropriate messages to the form elements
  validate(t, e, r, a) {
    var f;
    const { el: i, state: s, cancelled: o } = t.detail, c = (f = s == null ? void 0 : s[e]) == null ? void 0 : f.value;
    if (window.scrollTo({ top: 0, behavior: "smooth" }), o)
      return [!0, c];
    for (const h of r) {
      const g = h(c);
      if (V(this, N, G).call(this, i, e, g, a), g)
        return [!1, ""];
    }
    return [!0, c];
  }
  /**
   * Validates a group of fields ensuring that at least `minPassCount` of the items within the group
   * passes. This is useful in the scenario when n number fields are required out of n+m number of fields.
   *
   * @param {string[]} fields - An array of field names to be validated.
   * @param {Event} e - The event object associated with the validation trigger.
   * @param {FieldValidator[]} validators - An array of validator functions to apply to the fields.
   * @return {[number, Record<string, boolean>]} - Returns back the number of fields that passed and a record of the fields and their pass status.
   */
  validateGroup(t, e, r) {
    let a = 0;
    const i = {};
    for (const s of e) {
      const [o] = this.validate(t, s, r, { grouped: !0 });
      o && (i[s] = !0, a++);
    }
    return [a, i];
  }
  edit(t) {
    M(this._formRef, "external::alter:state", { index: t, operation: "edit" });
  }
  remove(t) {
    M(this._formRef, "external::alter:state", {
      index: t,
      operation: "remove"
    });
  }
  /**
   * Completes the form and triggers the onComplete callback.
   * This method should be used when you want to complete a form without navigating to a summary page.
   *
   * @important Developers must validate the form before calling this method.
   *
   * @example
   * // Validate first, then complete
   * const [isValid] = this.validate(e, "firstName", [
   *   requiredValidator("First name is required.")
   * ]);
   * if (isValid) {
   *   this.complete();
   * }
   * @returns void
   */
  complete() {
    if (!this._formRef) {
      console.error("complete: form ref is not set");
      return;
    }
    if (this._isCompleting) {
      console.warn("complete: completion already in progress");
      return;
    }
    this._isCompleting = !0, M(this._formRef, "fieldset::submit", null, { bubbles: !0 }), this._isCompleting = !1;
  }
  /**
   * Completes a subform and returns control to the parent form.
   * This method should be used when working with subforms that need to complete without a summary page.
   *
   * @important Developers must validate the subform before calling this method.
   *
   * @example
   * // Validate first, then complete the subform
   * const [isValid] = this._childFormController.validate(e, "fullName", [
   *   requiredValidator("Please enter the dependent's full name.")
   * ]);
   * if (isValid) {
   *   this._childFormController.completeSubform();
   * }
   * @returns void
   */
  completeSubform() {
    if (!this._formRef) {
      console.error("completeSubform: form ref is not set");
      return;
    }
    if (this._isCompleting) {
      console.warn("completeSubform: completion already in progress");
      return;
    }
    const t = this._formRef;
    this._isCompleting = !0;
    const e = (r) => {
      t.removeEventListener("_stateChange", e), L(t, "_complete", {}, { bubbles: !0 }), this._isCompleting = !1;
    };
    t.addEventListener("_stateChange", e), L(t, "_continue", null, { bubbles: !0 });
  }
  // removes any data collected that doesn't correspond with the final history path
  clean(t) {
    return t.history.reduce((e, r) => (e[r] = t.form[r], e), {});
  }
}
N = new WeakSet(), tt = function(t) {
  var e;
  Array.isArray(t.data) && (Array.isArray(this.state) || (this.state = {
    ...this.state,
    form: {
      ...((e = this.state) == null ? void 0 : e.form) || {},
      [t.id]: t.data
    }
  }));
}, // Private method to dispatch the error message to the form element
G = function(t, e, r, a) {
  t.dispatchEvent(
    new CustomEvent("msg", {
      composed: !0,
      detail: {
        action: "external::set:error",
        data: {
          name: e,
          msg: r,
          grouped: a == null ? void 0 : a.grouped
        }
      }
    })
  );
};
const et = ["basic", "success", "failure"];
function nt(n, t) {
  const e = crypto.randomUUID();
  return t = { uuid: e, type: "basic", ...t || {} }, !t.duration && t.type && et.includes(t.type) && (t.duration = "short"), M(
    document.body,
    "goa:temp-notification",
    { message: n, ...t },
    { bubbles: !0 }
  ), e;
}
function rt(n) {
  M(
    document.body,
    "goa:temp-notification:dismiss",
    n,
    { bubbles: !0 }
  );
}
function at(n, t) {
  M(
    document.body,
    "goa:temp-notification:progress",
    { uuid: n, progress: t },
    { bubbles: !0 }
  );
}
const Fe = {
  show: nt,
  dismiss: rt,
  setProgress: at
};
class it {
  constructor() {
    w(this, "done");
    this.done = /* @__PURE__ */ new Set();
  }
  when(t) {
    return t ? this : new st();
  }
  do(t, e) {
    this.done.has(t) || (this.done.add(t), e());
  }
}
class st extends it {
  do(t, e) {
  }
}
function d(n) {
  const t = Object.prototype.toString.call(n);
  return n instanceof Date || typeof n == "object" && t === "[object Date]" ? new n.constructor(+n) : typeof n == "number" || t === "[object Number]" || typeof n == "string" || t === "[object String]" ? new Date(n) : /* @__PURE__ */ new Date(NaN);
}
function y(n, t) {
  return n instanceof Date ? new n.constructor(t) : new Date(t);
}
function ot(n, t) {
  const e = d(n);
  return isNaN(t) ? y(n, NaN) : (t && e.setDate(e.getDate() + t), e);
}
function ut(n, t) {
  const e = d(n);
  if (isNaN(t)) return y(n, NaN);
  if (!t)
    return e;
  const r = e.getDate(), a = y(n, e.getTime());
  a.setMonth(e.getMonth() + t + 1, 0);
  const i = a.getDate();
  return r >= i ? a : (e.setFullYear(
    a.getFullYear(),
    a.getMonth(),
    r
  ), e);
}
const B = 6048e5, ct = 864e5;
let dt = {};
function W() {
  return dt;
}
function k(n, t) {
  var o, c, f, h;
  const e = W(), r = (t == null ? void 0 : t.weekStartsOn) ?? ((c = (o = t == null ? void 0 : t.locale) == null ? void 0 : o.options) == null ? void 0 : c.weekStartsOn) ?? e.weekStartsOn ?? ((h = (f = e.locale) == null ? void 0 : f.options) == null ? void 0 : h.weekStartsOn) ?? 0, a = d(n), i = a.getDay(), s = (i < r ? 7 : 0) + i - r;
  return a.setDate(a.getDate() - s), a.setHours(0, 0, 0, 0), a;
}
function T(n) {
  return k(n, { weekStartsOn: 1 });
}
function Q(n) {
  const t = d(n), e = t.getFullYear(), r = y(n, 0);
  r.setFullYear(e + 1, 0, 4), r.setHours(0, 0, 0, 0);
  const a = T(r), i = y(n, 0);
  i.setFullYear(e, 0, 4), i.setHours(0, 0, 0, 0);
  const s = T(i);
  return t.getTime() >= a.getTime() ? e + 1 : t.getTime() >= s.getTime() ? e : e - 1;
}
function Y(n) {
  const t = d(n);
  return t.setHours(0, 0, 0, 0), t;
}
function H(n) {
  const t = d(n), e = new Date(
    Date.UTC(
      t.getFullYear(),
      t.getMonth(),
      t.getDate(),
      t.getHours(),
      t.getMinutes(),
      t.getSeconds(),
      t.getMilliseconds()
    )
  );
  return e.setUTCFullYear(t.getFullYear()), +n - +e;
}
function ft(n, t) {
  const e = Y(n), r = Y(t), a = +e - H(e), i = +r - H(r);
  return Math.round((a - i) / ct);
}
function ht(n) {
  const t = Q(n), e = y(n, 0);
  return e.setFullYear(t, 0, 4), e.setHours(0, 0, 0, 0), T(e);
}
function lt(n, t) {
  const e = Y(n), r = Y(t);
  return +e == +r;
}
function mt(n) {
  return n instanceof Date || typeof n == "object" && Object.prototype.toString.call(n) === "[object Date]";
}
function gt(n) {
  if (!mt(n) && typeof n != "number")
    return !1;
  const t = d(n);
  return !isNaN(Number(t));
}
function yt(n) {
  const t = d(n), e = y(n, 0);
  return e.setFullYear(t.getFullYear(), 0, 1), e.setHours(0, 0, 0, 0), e;
}
const bt = {
  lessThanXSeconds: {
    one: "less than a second",
    other: "less than {{count}} seconds"
  },
  xSeconds: {
    one: "1 second",
    other: "{{count}} seconds"
  },
  halfAMinute: "half a minute",
  lessThanXMinutes: {
    one: "less than a minute",
    other: "less than {{count}} minutes"
  },
  xMinutes: {
    one: "1 minute",
    other: "{{count}} minutes"
  },
  aboutXHours: {
    one: "about 1 hour",
    other: "about {{count}} hours"
  },
  xHours: {
    one: "1 hour",
    other: "{{count}} hours"
  },
  xDays: {
    one: "1 day",
    other: "{{count}} days"
  },
  aboutXWeeks: {
    one: "about 1 week",
    other: "about {{count}} weeks"
  },
  xWeeks: {
    one: "1 week",
    other: "{{count}} weeks"
  },
  aboutXMonths: {
    one: "about 1 month",
    other: "about {{count}} months"
  },
  xMonths: {
    one: "1 month",
    other: "{{count}} months"
  },
  aboutXYears: {
    one: "about 1 year",
    other: "about {{count}} years"
  },
  xYears: {
    one: "1 year",
    other: "{{count}} years"
  },
  overXYears: {
    one: "over 1 year",
    other: "over {{count}} years"
  },
  almostXYears: {
    one: "almost 1 year",
    other: "almost {{count}} years"
  }
}, wt = (n, t, e) => {
  let r;
  const a = bt[n];
  return typeof a == "string" ? r = a : t === 1 ? r = a.one : r = a.other.replace("{{count}}", t.toString()), e != null && e.addSuffix ? e.comparison && e.comparison > 0 ? "in " + r : r + " ago" : r;
};
function E(n) {
  return (t = {}) => {
    const e = t.width ? String(t.width) : n.defaultWidth;
    return n.formats[e] || n.formats[n.defaultWidth];
  };
}
const pt = {
  full: "EEEE, MMMM do, y",
  long: "MMMM do, y",
  medium: "MMM d, y",
  short: "MM/dd/yyyy"
}, Mt = {
  full: "h:mm:ss a zzzz",
  long: "h:mm:ss a z",
  medium: "h:mm:ss a",
  short: "h:mm a"
}, Dt = {
  full: "{{date}} 'at' {{time}}",
  long: "{{date}} 'at' {{time}}",
  medium: "{{date}}, {{time}}",
  short: "{{date}}, {{time}}"
}, Ot = {
  date: E({
    formats: pt,
    defaultWidth: "full"
  }),
  time: E({
    formats: Mt,
    defaultWidth: "full"
  }),
  dateTime: E({
    formats: Dt,
    defaultWidth: "full"
  })
}, St = {
  lastWeek: "'last' eeee 'at' p",
  yesterday: "'yesterday at' p",
  today: "'today at' p",
  tomorrow: "'tomorrow at' p",
  nextWeek: "eeee 'at' p",
  other: "P"
}, Pt = (n, t, e, r) => St[n];
function v(n) {
  return (t, e) => {
    const r = e != null && e.context ? String(e.context) : "standalone";
    let a;
    if (r === "formatting" && n.formattingValues) {
      const s = n.defaultFormattingWidth || n.defaultWidth, o = e != null && e.width ? String(e.width) : s;
      a = n.formattingValues[o] || n.formattingValues[s];
    } else {
      const s = n.defaultWidth, o = e != null && e.width ? String(e.width) : n.defaultWidth;
      a = n.values[o] || n.values[s];
    }
    const i = n.argumentCallback ? n.argumentCallback(t) : t;
    return a[i];
  };
}
const vt = {
  narrow: ["B", "A"],
  abbreviated: ["BC", "AD"],
  wide: ["Before Christ", "Anno Domini"]
}, xt = {
  narrow: ["1", "2", "3", "4"],
  abbreviated: ["Q1", "Q2", "Q3", "Q4"],
  wide: ["1st quarter", "2nd quarter", "3rd quarter", "4th quarter"]
}, kt = {
  narrow: ["J", "F", "M", "A", "M", "J", "J", "A", "S", "O", "N", "D"],
  abbreviated: [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec"
  ],
  wide: [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December"
  ]
}, Nt = {
  narrow: ["S", "M", "T", "W", "T", "F", "S"],
  short: ["Su", "Mo", "Tu", "We", "Th", "Fr", "Sa"],
  abbreviated: ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
  wide: [
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday"
  ]
}, Tt = {
  narrow: {
    am: "a",
    pm: "p",
    midnight: "mi",
    noon: "n",
    morning: "morning",
    afternoon: "afternoon",
    evening: "evening",
    night: "night"
  },
  abbreviated: {
    am: "AM",
    pm: "PM",
    midnight: "midnight",
    noon: "noon",
    morning: "morning",
    afternoon: "afternoon",
    evening: "evening",
    night: "night"
  },
  wide: {
    am: "a.m.",
    pm: "p.m.",
    midnight: "midnight",
    noon: "noon",
    morning: "morning",
    afternoon: "afternoon",
    evening: "evening",
    night: "night"
  }
}, Yt = {
  narrow: {
    am: "a",
    pm: "p",
    midnight: "mi",
    noon: "n",
    morning: "in the morning",
    afternoon: "in the afternoon",
    evening: "in the evening",
    night: "at night"
  },
  abbreviated: {
    am: "AM",
    pm: "PM",
    midnight: "midnight",
    noon: "noon",
    morning: "in the morning",
    afternoon: "in the afternoon",
    evening: "in the evening",
    night: "at night"
  },
  wide: {
    am: "a.m.",
    pm: "p.m.",
    midnight: "midnight",
    noon: "noon",
    morning: "in the morning",
    afternoon: "in the afternoon",
    evening: "in the evening",
    night: "at night"
  }
}, Wt = (n, t) => {
  const e = Number(n), r = e % 100;
  if (r > 20 || r < 10)
    switch (r % 10) {
      case 1:
        return e + "st";
      case 2:
        return e + "nd";
      case 3:
        return e + "rd";
    }
  return e + "th";
}, _t = {
  ordinalNumber: Wt,
  era: v({
    values: vt,
    defaultWidth: "wide"
  }),
  quarter: v({
    values: xt,
    defaultWidth: "wide",
    argumentCallback: (n) => n - 1
  }),
  month: v({
    values: kt,
    defaultWidth: "wide"
  }),
  day: v({
    values: Nt,
    defaultWidth: "wide"
  }),
  dayPeriod: v({
    values: Tt,
    defaultWidth: "wide",
    formattingValues: Yt,
    defaultFormattingWidth: "wide"
  })
};
function x(n) {
  return (t, e = {}) => {
    const r = e.width, a = r && n.matchPatterns[r] || n.matchPatterns[n.defaultMatchWidth], i = t.match(a);
    if (!i)
      return null;
    const s = i[0], o = r && n.parsePatterns[r] || n.parsePatterns[n.defaultParseWidth], c = Array.isArray(o) ? Ft(o, (g) => g.test(s)) : (
      // eslint-disable-next-line @typescript-eslint/no-explicit-any -- I challange you to fix the type
      Et(o, (g) => g.test(s))
    );
    let f;
    f = n.valueCallback ? n.valueCallback(c) : c, f = e.valueCallback ? (
      // eslint-disable-next-line @typescript-eslint/no-explicit-any -- I challange you to fix the type
      e.valueCallback(f)
    ) : f;
    const h = t.slice(s.length);
    return { value: f, rest: h };
  };
}
function Et(n, t) {
  for (const e in n)
    if (Object.prototype.hasOwnProperty.call(n, e) && t(n[e]))
      return e;
}
function Ft(n, t) {
  for (let e = 0; e < n.length; e++)
    if (t(n[e]))
      return e;
}
function Ct(n) {
  return (t, e = {}) => {
    const r = t.match(n.matchPattern);
    if (!r) return null;
    const a = r[0], i = t.match(n.parsePattern);
    if (!i) return null;
    let s = n.valueCallback ? n.valueCallback(i[0]) : i[0];
    s = e.valueCallback ? e.valueCallback(s) : s;
    const o = t.slice(a.length);
    return { value: s, rest: o };
  };
}
const At = /^(\d+)(th|st|nd|rd)?/i, qt = /\d+/i, Rt = {
  narrow: /^(b|a)/i,
  abbreviated: /^(b\.?\s?c\.?|b\.?\s?c\.?\s?e\.?|a\.?\s?d\.?|c\.?\s?e\.?)/i,
  wide: /^(before christ|before common era|anno domini|common era)/i
}, Vt = {
  any: [/^b/i, /^(a|c)/i]
}, Lt = {
  narrow: /^[1234]/i,
  abbreviated: /^q[1234]/i,
  wide: /^[1234](th|st|nd|rd)? quarter/i
}, Ht = {
  any: [/1/i, /2/i, /3/i, /4/i]
}, It = {
  narrow: /^[jfmasond]/i,
  abbreviated: /^(jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)/i,
  wide: /^(january|february|march|april|may|june|july|august|september|october|november|december)/i
}, Xt = {
  narrow: [
    /^j/i,
    /^f/i,
    /^m/i,
    /^a/i,
    /^m/i,
    /^j/i,
    /^j/i,
    /^a/i,
    /^s/i,
    /^o/i,
    /^n/i,
    /^d/i
  ],
  any: [
    /^ja/i,
    /^f/i,
    /^mar/i,
    /^ap/i,
    /^may/i,
    /^jun/i,
    /^jul/i,
    /^au/i,
    /^s/i,
    /^o/i,
    /^n/i,
    /^d/i
  ]
}, jt = {
  narrow: /^[smtwf]/i,
  short: /^(su|mo|tu|we|th|fr|sa)/i,
  abbreviated: /^(sun|mon|tue|wed|thu|fri|sat)/i,
  wide: /^(sunday|monday|tuesday|wednesday|thursday|friday|saturday)/i
}, $t = {
  narrow: [/^s/i, /^m/i, /^t/i, /^w/i, /^t/i, /^f/i, /^s/i],
  any: [/^su/i, /^m/i, /^tu/i, /^w/i, /^th/i, /^f/i, /^sa/i]
}, Gt = {
  narrow: /^(a|p|mi|n|(in the|at) (morning|afternoon|evening|night))/i,
  any: /^([ap]\.?\s?m\.?|midnight|noon|(in the|at) (morning|afternoon|evening|night))/i
}, Bt = {
  any: {
    am: /^a/i,
    pm: /^p/i,
    midnight: /^mi/i,
    noon: /^no/i,
    morning: /morning/i,
    afternoon: /afternoon/i,
    evening: /evening/i,
    night: /night/i
  }
}, Qt = {
  ordinalNumber: Ct({
    matchPattern: At,
    parsePattern: qt,
    valueCallback: (n) => parseInt(n, 10)
  }),
  era: x({
    matchPatterns: Rt,
    defaultMatchWidth: "wide",
    parsePatterns: Vt,
    defaultParseWidth: "any"
  }),
  quarter: x({
    matchPatterns: Lt,
    defaultMatchWidth: "wide",
    parsePatterns: Ht,
    defaultParseWidth: "any",
    valueCallback: (n) => n + 1
  }),
  month: x({
    matchPatterns: It,
    defaultMatchWidth: "wide",
    parsePatterns: Xt,
    defaultParseWidth: "any"
  }),
  day: x({
    matchPatterns: jt,
    defaultMatchWidth: "wide",
    parsePatterns: $t,
    defaultParseWidth: "any"
  }),
  dayPeriod: x({
    matchPatterns: Gt,
    defaultMatchWidth: "any",
    parsePatterns: Bt,
    defaultParseWidth: "any"
  })
}, Jt = {
  code: "en-US",
  formatDistance: wt,
  formatLong: Ot,
  formatRelative: Pt,
  localize: _t,
  match: Qt,
  options: {
    weekStartsOn: 0,
    firstWeekContainsDate: 1
  }
};
function Ut(n) {
  const t = d(n);
  return ft(t, yt(t)) + 1;
}
function Zt(n) {
  const t = d(n), e = +T(t) - +ht(t);
  return Math.round(e / B) + 1;
}
function J(n, t) {
  var h, g, S, P;
  const e = d(n), r = e.getFullYear(), a = W(), i = (t == null ? void 0 : t.firstWeekContainsDate) ?? ((g = (h = t == null ? void 0 : t.locale) == null ? void 0 : h.options) == null ? void 0 : g.firstWeekContainsDate) ?? a.firstWeekContainsDate ?? ((P = (S = a.locale) == null ? void 0 : S.options) == null ? void 0 : P.firstWeekContainsDate) ?? 1, s = y(n, 0);
  s.setFullYear(r + 1, 0, i), s.setHours(0, 0, 0, 0);
  const o = k(s, t), c = y(n, 0);
  c.setFullYear(r, 0, i), c.setHours(0, 0, 0, 0);
  const f = k(c, t);
  return e.getTime() >= o.getTime() ? r + 1 : e.getTime() >= f.getTime() ? r : r - 1;
}
function zt(n, t) {
  var o, c, f, h;
  const e = W(), r = (t == null ? void 0 : t.firstWeekContainsDate) ?? ((c = (o = t == null ? void 0 : t.locale) == null ? void 0 : o.options) == null ? void 0 : c.firstWeekContainsDate) ?? e.firstWeekContainsDate ?? ((h = (f = e.locale) == null ? void 0 : f.options) == null ? void 0 : h.firstWeekContainsDate) ?? 1, a = J(n, t), i = y(n, 0);
  return i.setFullYear(a, 0, r), i.setHours(0, 0, 0, 0), k(i, t);
}
function Kt(n, t) {
  const e = d(n), r = +k(e, t) - +zt(e, t);
  return Math.round(r / B) + 1;
}
function u(n, t) {
  const e = n < 0 ? "-" : "", r = Math.abs(n).toString().padStart(t, "0");
  return e + r;
}
const p = {
  // Year
  y(n, t) {
    const e = n.getFullYear(), r = e > 0 ? e : 1 - e;
    return u(t === "yy" ? r % 100 : r, t.length);
  },
  // Month
  M(n, t) {
    const e = n.getMonth();
    return t === "M" ? String(e + 1) : u(e + 1, 2);
  },
  // Day of the month
  d(n, t) {
    return u(n.getDate(), t.length);
  },
  // AM or PM
  a(n, t) {
    const e = n.getHours() / 12 >= 1 ? "pm" : "am";
    switch (t) {
      case "a":
      case "aa":
        return e.toUpperCase();
      case "aaa":
        return e;
      case "aaaaa":
        return e[0];
      case "aaaa":
      default:
        return e === "am" ? "a.m." : "p.m.";
    }
  },
  // Hour [1-12]
  h(n, t) {
    return u(n.getHours() % 12 || 12, t.length);
  },
  // Hour [0-23]
  H(n, t) {
    return u(n.getHours(), t.length);
  },
  // Minute
  m(n, t) {
    return u(n.getMinutes(), t.length);
  },
  // Second
  s(n, t) {
    return u(n.getSeconds(), t.length);
  },
  // Fraction of second
  S(n, t) {
    const e = t.length, r = n.getMilliseconds(), a = Math.trunc(
      r * Math.pow(10, e - 3)
    );
    return u(a, t.length);
  }
}, O = {
  midnight: "midnight",
  noon: "noon",
  morning: "morning",
  afternoon: "afternoon",
  evening: "evening",
  night: "night"
}, I = {
  // Era
  G: function(n, t, e) {
    const r = n.getFullYear() > 0 ? 1 : 0;
    switch (t) {
      // AD, BC
      case "G":
      case "GG":
      case "GGG":
        return e.era(r, { width: "abbreviated" });
      // A, B
      case "GGGGG":
        return e.era(r, { width: "narrow" });
      // Anno Domini, Before Christ
      case "GGGG":
      default:
        return e.era(r, { width: "wide" });
    }
  },
  // Year
  y: function(n, t, e) {
    if (t === "yo") {
      const r = n.getFullYear(), a = r > 0 ? r : 1 - r;
      return e.ordinalNumber(a, { unit: "year" });
    }
    return p.y(n, t);
  },
  // Local week-numbering year
  Y: function(n, t, e, r) {
    const a = J(n, r), i = a > 0 ? a : 1 - a;
    if (t === "YY") {
      const s = i % 100;
      return u(s, 2);
    }
    return t === "Yo" ? e.ordinalNumber(i, { unit: "year" }) : u(i, t.length);
  },
  // ISO week-numbering year
  R: function(n, t) {
    const e = Q(n);
    return u(e, t.length);
  },
  // Extended year. This is a single number designating the year of this calendar system.
  // The main difference between `y` and `u` localizers are B.C. years:
  // | Year | `y` | `u` |
  // |------|-----|-----|
  // | AC 1 |   1 |   1 |
  // | BC 1 |   1 |   0 |
  // | BC 2 |   2 |  -1 |
  // Also `yy` always returns the last two digits of a year,
  // while `uu` pads single digit years to 2 characters and returns other years unchanged.
  u: function(n, t) {
    const e = n.getFullYear();
    return u(e, t.length);
  },
  // Quarter
  Q: function(n, t, e) {
    const r = Math.ceil((n.getMonth() + 1) / 3);
    switch (t) {
      // 1, 2, 3, 4
      case "Q":
        return String(r);
      // 01, 02, 03, 04
      case "QQ":
        return u(r, 2);
      // 1st, 2nd, 3rd, 4th
      case "Qo":
        return e.ordinalNumber(r, { unit: "quarter" });
      // Q1, Q2, Q3, Q4
      case "QQQ":
        return e.quarter(r, {
          width: "abbreviated",
          context: "formatting"
        });
      // 1, 2, 3, 4 (narrow quarter; could be not numerical)
      case "QQQQQ":
        return e.quarter(r, {
          width: "narrow",
          context: "formatting"
        });
      // 1st quarter, 2nd quarter, ...
      case "QQQQ":
      default:
        return e.quarter(r, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // Stand-alone quarter
  q: function(n, t, e) {
    const r = Math.ceil((n.getMonth() + 1) / 3);
    switch (t) {
      // 1, 2, 3, 4
      case "q":
        return String(r);
      // 01, 02, 03, 04
      case "qq":
        return u(r, 2);
      // 1st, 2nd, 3rd, 4th
      case "qo":
        return e.ordinalNumber(r, { unit: "quarter" });
      // Q1, Q2, Q3, Q4
      case "qqq":
        return e.quarter(r, {
          width: "abbreviated",
          context: "standalone"
        });
      // 1, 2, 3, 4 (narrow quarter; could be not numerical)
      case "qqqqq":
        return e.quarter(r, {
          width: "narrow",
          context: "standalone"
        });
      // 1st quarter, 2nd quarter, ...
      case "qqqq":
      default:
        return e.quarter(r, {
          width: "wide",
          context: "standalone"
        });
    }
  },
  // Month
  M: function(n, t, e) {
    const r = n.getMonth();
    switch (t) {
      case "M":
      case "MM":
        return p.M(n, t);
      // 1st, 2nd, ..., 12th
      case "Mo":
        return e.ordinalNumber(r + 1, { unit: "month" });
      // Jan, Feb, ..., Dec
      case "MMM":
        return e.month(r, {
          width: "abbreviated",
          context: "formatting"
        });
      // J, F, ..., D
      case "MMMMM":
        return e.month(r, {
          width: "narrow",
          context: "formatting"
        });
      // January, February, ..., December
      case "MMMM":
      default:
        return e.month(r, { width: "wide", context: "formatting" });
    }
  },
  // Stand-alone month
  L: function(n, t, e) {
    const r = n.getMonth();
    switch (t) {
      // 1, 2, ..., 12
      case "L":
        return String(r + 1);
      // 01, 02, ..., 12
      case "LL":
        return u(r + 1, 2);
      // 1st, 2nd, ..., 12th
      case "Lo":
        return e.ordinalNumber(r + 1, { unit: "month" });
      // Jan, Feb, ..., Dec
      case "LLL":
        return e.month(r, {
          width: "abbreviated",
          context: "standalone"
        });
      // J, F, ..., D
      case "LLLLL":
        return e.month(r, {
          width: "narrow",
          context: "standalone"
        });
      // January, February, ..., December
      case "LLLL":
      default:
        return e.month(r, { width: "wide", context: "standalone" });
    }
  },
  // Local week of year
  w: function(n, t, e, r) {
    const a = Kt(n, r);
    return t === "wo" ? e.ordinalNumber(a, { unit: "week" }) : u(a, t.length);
  },
  // ISO week of year
  I: function(n, t, e) {
    const r = Zt(n);
    return t === "Io" ? e.ordinalNumber(r, { unit: "week" }) : u(r, t.length);
  },
  // Day of the month
  d: function(n, t, e) {
    return t === "do" ? e.ordinalNumber(n.getDate(), { unit: "date" }) : p.d(n, t);
  },
  // Day of year
  D: function(n, t, e) {
    const r = Ut(n);
    return t === "Do" ? e.ordinalNumber(r, { unit: "dayOfYear" }) : u(r, t.length);
  },
  // Day of week
  E: function(n, t, e) {
    const r = n.getDay();
    switch (t) {
      // Tue
      case "E":
      case "EE":
      case "EEE":
        return e.day(r, {
          width: "abbreviated",
          context: "formatting"
        });
      // T
      case "EEEEE":
        return e.day(r, {
          width: "narrow",
          context: "formatting"
        });
      // Tu
      case "EEEEEE":
        return e.day(r, {
          width: "short",
          context: "formatting"
        });
      // Tuesday
      case "EEEE":
      default:
        return e.day(r, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // Local day of week
  e: function(n, t, e, r) {
    const a = n.getDay(), i = (a - r.weekStartsOn + 8) % 7 || 7;
    switch (t) {
      // Numerical value (Nth day of week with current locale or weekStartsOn)
      case "e":
        return String(i);
      // Padded numerical value
      case "ee":
        return u(i, 2);
      // 1st, 2nd, ..., 7th
      case "eo":
        return e.ordinalNumber(i, { unit: "day" });
      case "eee":
        return e.day(a, {
          width: "abbreviated",
          context: "formatting"
        });
      // T
      case "eeeee":
        return e.day(a, {
          width: "narrow",
          context: "formatting"
        });
      // Tu
      case "eeeeee":
        return e.day(a, {
          width: "short",
          context: "formatting"
        });
      // Tuesday
      case "eeee":
      default:
        return e.day(a, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // Stand-alone local day of week
  c: function(n, t, e, r) {
    const a = n.getDay(), i = (a - r.weekStartsOn + 8) % 7 || 7;
    switch (t) {
      // Numerical value (same as in `e`)
      case "c":
        return String(i);
      // Padded numerical value
      case "cc":
        return u(i, t.length);
      // 1st, 2nd, ..., 7th
      case "co":
        return e.ordinalNumber(i, { unit: "day" });
      case "ccc":
        return e.day(a, {
          width: "abbreviated",
          context: "standalone"
        });
      // T
      case "ccccc":
        return e.day(a, {
          width: "narrow",
          context: "standalone"
        });
      // Tu
      case "cccccc":
        return e.day(a, {
          width: "short",
          context: "standalone"
        });
      // Tuesday
      case "cccc":
      default:
        return e.day(a, {
          width: "wide",
          context: "standalone"
        });
    }
  },
  // ISO day of week
  i: function(n, t, e) {
    const r = n.getDay(), a = r === 0 ? 7 : r;
    switch (t) {
      // 2
      case "i":
        return String(a);
      // 02
      case "ii":
        return u(a, t.length);
      // 2nd
      case "io":
        return e.ordinalNumber(a, { unit: "day" });
      // Tue
      case "iii":
        return e.day(r, {
          width: "abbreviated",
          context: "formatting"
        });
      // T
      case "iiiii":
        return e.day(r, {
          width: "narrow",
          context: "formatting"
        });
      // Tu
      case "iiiiii":
        return e.day(r, {
          width: "short",
          context: "formatting"
        });
      // Tuesday
      case "iiii":
      default:
        return e.day(r, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // AM or PM
  a: function(n, t, e) {
    const a = n.getHours() / 12 >= 1 ? "pm" : "am";
    switch (t) {
      case "a":
      case "aa":
        return e.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        });
      case "aaa":
        return e.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        }).toLowerCase();
      case "aaaaa":
        return e.dayPeriod(a, {
          width: "narrow",
          context: "formatting"
        });
      case "aaaa":
      default:
        return e.dayPeriod(a, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // AM, PM, midnight, noon
  b: function(n, t, e) {
    const r = n.getHours();
    let a;
    switch (r === 12 ? a = O.noon : r === 0 ? a = O.midnight : a = r / 12 >= 1 ? "pm" : "am", t) {
      case "b":
      case "bb":
        return e.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        });
      case "bbb":
        return e.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        }).toLowerCase();
      case "bbbbb":
        return e.dayPeriod(a, {
          width: "narrow",
          context: "formatting"
        });
      case "bbbb":
      default:
        return e.dayPeriod(a, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // in the morning, in the afternoon, in the evening, at night
  B: function(n, t, e) {
    const r = n.getHours();
    let a;
    switch (r >= 17 ? a = O.evening : r >= 12 ? a = O.afternoon : r >= 4 ? a = O.morning : a = O.night, t) {
      case "B":
      case "BB":
      case "BBB":
        return e.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        });
      case "BBBBB":
        return e.dayPeriod(a, {
          width: "narrow",
          context: "formatting"
        });
      case "BBBB":
      default:
        return e.dayPeriod(a, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // Hour [1-12]
  h: function(n, t, e) {
    if (t === "ho") {
      let r = n.getHours() % 12;
      return r === 0 && (r = 12), e.ordinalNumber(r, { unit: "hour" });
    }
    return p.h(n, t);
  },
  // Hour [0-23]
  H: function(n, t, e) {
    return t === "Ho" ? e.ordinalNumber(n.getHours(), { unit: "hour" }) : p.H(n, t);
  },
  // Hour [0-11]
  K: function(n, t, e) {
    const r = n.getHours() % 12;
    return t === "Ko" ? e.ordinalNumber(r, { unit: "hour" }) : u(r, t.length);
  },
  // Hour [1-24]
  k: function(n, t, e) {
    let r = n.getHours();
    return r === 0 && (r = 24), t === "ko" ? e.ordinalNumber(r, { unit: "hour" }) : u(r, t.length);
  },
  // Minute
  m: function(n, t, e) {
    return t === "mo" ? e.ordinalNumber(n.getMinutes(), { unit: "minute" }) : p.m(n, t);
  },
  // Second
  s: function(n, t, e) {
    return t === "so" ? e.ordinalNumber(n.getSeconds(), { unit: "second" }) : p.s(n, t);
  },
  // Fraction of second
  S: function(n, t) {
    return p.S(n, t);
  },
  // Timezone (ISO-8601. If offset is 0, output is always `'Z'`)
  X: function(n, t, e) {
    const r = n.getTimezoneOffset();
    if (r === 0)
      return "Z";
    switch (t) {
      // Hours and optional minutes
      case "X":
        return j(r);
      // Hours, minutes and optional seconds without `:` delimiter
      // Note: neither ISO-8601 nor JavaScript supports seconds in timezone offsets
      // so this token always has the same output as `XX`
      case "XXXX":
      case "XX":
        return D(r);
      // Hours, minutes and optional seconds with `:` delimiter
      // Note: neither ISO-8601 nor JavaScript supports seconds in timezone offsets
      // so this token always has the same output as `XXX`
      case "XXXXX":
      case "XXX":
      // Hours and minutes with `:` delimiter
      default:
        return D(r, ":");
    }
  },
  // Timezone (ISO-8601. If offset is 0, output is `'+00:00'` or equivalent)
  x: function(n, t, e) {
    const r = n.getTimezoneOffset();
    switch (t) {
      // Hours and optional minutes
      case "x":
        return j(r);
      // Hours, minutes and optional seconds without `:` delimiter
      // Note: neither ISO-8601 nor JavaScript supports seconds in timezone offsets
      // so this token always has the same output as `xx`
      case "xxxx":
      case "xx":
        return D(r);
      // Hours, minutes and optional seconds with `:` delimiter
      // Note: neither ISO-8601 nor JavaScript supports seconds in timezone offsets
      // so this token always has the same output as `xxx`
      case "xxxxx":
      case "xxx":
      // Hours and minutes with `:` delimiter
      default:
        return D(r, ":");
    }
  },
  // Timezone (GMT)
  O: function(n, t, e) {
    const r = n.getTimezoneOffset();
    switch (t) {
      // Short
      case "O":
      case "OO":
      case "OOO":
        return "GMT" + X(r, ":");
      // Long
      case "OOOO":
      default:
        return "GMT" + D(r, ":");
    }
  },
  // Timezone (specific non-location)
  z: function(n, t, e) {
    const r = n.getTimezoneOffset();
    switch (t) {
      // Short
      case "z":
      case "zz":
      case "zzz":
        return "GMT" + X(r, ":");
      // Long
      case "zzzz":
      default:
        return "GMT" + D(r, ":");
    }
  },
  // Seconds timestamp
  t: function(n, t, e) {
    const r = Math.trunc(n.getTime() / 1e3);
    return u(r, t.length);
  },
  // Milliseconds timestamp
  T: function(n, t, e) {
    const r = n.getTime();
    return u(r, t.length);
  }
};
function X(n, t = "") {
  const e = n > 0 ? "-" : "+", r = Math.abs(n), a = Math.trunc(r / 60), i = r % 60;
  return i === 0 ? e + String(a) : e + String(a) + t + u(i, 2);
}
function j(n, t) {
  return n % 60 === 0 ? (n > 0 ? "-" : "+") + u(Math.abs(n) / 60, 2) : D(n, t);
}
function D(n, t = "") {
  const e = n > 0 ? "-" : "+", r = Math.abs(n), a = u(Math.trunc(r / 60), 2), i = u(r % 60, 2);
  return e + a + t + i;
}
const $ = (n, t) => {
  switch (n) {
    case "P":
      return t.date({ width: "short" });
    case "PP":
      return t.date({ width: "medium" });
    case "PPP":
      return t.date({ width: "long" });
    case "PPPP":
    default:
      return t.date({ width: "full" });
  }
}, U = (n, t) => {
  switch (n) {
    case "p":
      return t.time({ width: "short" });
    case "pp":
      return t.time({ width: "medium" });
    case "ppp":
      return t.time({ width: "long" });
    case "pppp":
    default:
      return t.time({ width: "full" });
  }
}, te = (n, t) => {
  const e = n.match(/(P+)(p+)?/) || [], r = e[1], a = e[2];
  if (!a)
    return $(n, t);
  let i;
  switch (r) {
    case "P":
      i = t.dateTime({ width: "short" });
      break;
    case "PP":
      i = t.dateTime({ width: "medium" });
      break;
    case "PPP":
      i = t.dateTime({ width: "long" });
      break;
    case "PPPP":
    default:
      i = t.dateTime({ width: "full" });
      break;
  }
  return i.replace("{{date}}", $(r, t)).replace("{{time}}", U(a, t));
}, ee = {
  p: U,
  P: te
}, ne = /^D+$/, re = /^Y+$/, ae = ["D", "DD", "YY", "YYYY"];
function ie(n) {
  return ne.test(n);
}
function se(n) {
  return re.test(n);
}
function oe(n, t, e) {
  const r = ue(n, t, e);
  if (console.warn(r), ae.includes(n)) throw new RangeError(r);
}
function ue(n, t, e) {
  const r = n[0] === "Y" ? "years" : "days of the month";
  return `Use \`${n.toLowerCase()}\` instead of \`${n}\` (in \`${t}\`) for formatting ${r} to the input \`${e}\`; see: https://github.com/date-fns/date-fns/blob/master/docs/unicodeTokens.md`;
}
const ce = /[yYQqMLwIdDecihHKkms]o|(\w)\1*|''|'(''|[^'])+('|$)|./g, de = /P+p+|P+|p+|''|'(''|[^'])+('|$)|./g, fe = /^'([^]*?)'?$/, he = /''/g, le = /[a-zA-Z]/;
function me(n, t, e) {
  var h, g, S, P;
  const r = W(), a = r.locale ?? Jt, i = r.firstWeekContainsDate ?? ((g = (h = r.locale) == null ? void 0 : h.options) == null ? void 0 : g.firstWeekContainsDate) ?? 1, s = r.weekStartsOn ?? ((P = (S = r.locale) == null ? void 0 : S.options) == null ? void 0 : P.weekStartsOn) ?? 0, o = d(n);
  if (!gt(o))
    throw new RangeError("Invalid time value");
  let c = t.match(de).map((m) => {
    const l = m[0];
    if (l === "p" || l === "P") {
      const _ = ee[l];
      return _(m, a.formatLong);
    }
    return m;
  }).join("").match(ce).map((m) => {
    if (m === "''")
      return { isToken: !1, value: "'" };
    const l = m[0];
    if (l === "'")
      return { isToken: !1, value: ge(m) };
    if (I[l])
      return { isToken: !0, value: m };
    if (l.match(le))
      throw new RangeError(
        "Format string contains an unescaped latin alphabet character `" + l + "`"
      );
    return { isToken: !1, value: m };
  });
  a.localize.preprocessor && (c = a.localize.preprocessor(o, c));
  const f = {
    firstWeekContainsDate: i,
    weekStartsOn: s,
    locale: a
  };
  return c.map((m) => {
    if (!m.isToken) return m.value;
    const l = m.value;
    (se(l) || ie(l)) && oe(l, t, String(n));
    const _ = I[l[0]];
    return _(o, l, a.localize, f);
  }).join("");
}
function ge(n) {
  const t = n.match(fe);
  return t ? t[1].replace(he, "'") : n;
}
function ye(n) {
  const t = d(n), e = t.getFullYear(), r = t.getMonth(), a = y(n, 0);
  return a.setFullYear(e, r + 1, 0), a.setHours(0, 0, 0, 0), a.getDate();
}
function be(n) {
  const t = d(n), e = t.getMonth();
  return t.setFullYear(t.getFullYear(), e + 1, 0), t.setHours(0, 0, 0, 0), t;
}
function we(n, t) {
  const e = d(n), r = d(t);
  return e.getTime() > r.getTime();
}
function pe(n, t) {
  const e = d(n), r = d(t);
  return +e < +r;
}
function Me(n, t) {
  const e = d(n), r = d(t);
  return e.getFullYear() === r.getFullYear() && e.getMonth() === r.getMonth();
}
class b {
  constructor(t) {
    w(this, "_dateNums");
    t || t === 0 ? this._dateNums = b.parse(t) : this._dateNums = b.parse(/* @__PURE__ */ new Date());
  }
  static parse(t) {
    return typeof t == "string" ? (t = t.split("T")[0], t.split("-").map((e) => +e)) : t instanceof Date ? [t.getFullYear(), t.getMonth() + 1, t.getDate()] : t === 0 ? [0, 0, 0] : [t.year, t.month, t.day];
  }
  static init() {
    return new b(0);
  }
  // Used internally to get the date value for the date_fns
  get date() {
    return new Date(
      this._dateNums[0],
      this._dateNums[1] - 1,
      this._dateNums[2]
    );
  }
  get year() {
    return this._dateNums[0];
  }
  get month() {
    return this._dateNums[1];
  }
  get day() {
    return this._dateNums[2];
  }
  get dayOfWeek() {
    return this.date.getDay();
  }
  get daysInMonth() {
    return ye(this.date);
  }
  get firstDayOfMonth() {
    return new b({ year: this.year, month: this.month, day: 1 });
  }
  get lastDayOfMonth() {
    return new b(be(this.date));
  }
  get previousDay() {
    return this.clone().addDays(-1);
  }
  get nextDay() {
    return this.clone().addDays(1);
  }
  get previousWeek() {
    return this.clone().addDays(-7);
  }
  get nextWeek() {
    return this.clone().addDays(7);
  }
  get previousMonth() {
    return this.clone().addMonths(-1);
  }
  get nextMonth() {
    return this.clone().addMonths(1);
  }
  clone() {
    return new b(this.toString());
  }
  setYear(t) {
    this._dateNums[0] = t;
  }
  setMonth(t) {
    this._dateNums[1] = t;
  }
  setDay(t) {
    return this._dateNums[2] = t, this;
  }
  addYears(t) {
    return this._dateNums[0] += t, this;
  }
  addMonths(t) {
    return this._dateNums = b.parse(ut(this.date, t)), this;
  }
  addDays(t) {
    return this._dateNums = b.parse(ot(this.date, t)), this;
  }
  isSameDay(t) {
    return lt(this.date, t.date);
  }
  isSameMonth(t) {
    return Me(this.date, t.date);
  }
  isBefore(t) {
    return pe(this.date, t.date);
  }
  isAfter(t) {
    return we(this.date, t.date);
  }
  isZero() {
    return this._dateNums[0] === 0 && this._dateNums[1] === 0 && this._dateNums[2] === 0;
  }
  isValid() {
    const t = new Date(this.toString());
    return !(isNaN(t.getTime()) || this.toString() !== t.toISOString().split("T")[0]);
  }
  format(t) {
    return this.isZero() ? "" : me(this.date, t);
  }
  toString() {
    return this.isZero() ? "" : this._dateNums.map((t) => `${t}`.length < 2 ? `0${t}` : `${t}`).join("-");
  }
}
export {
  b as CalendarDate,
  Se as FormValidator,
  Oe as GoabWorkspaceLayoutScrollState,
  it as Once,
  Ee as PublicFormController,
  Te as SINValidator,
  Fe as TemporaryNotification,
  Pe as birthDayValidator,
  ve as birthMonthValidator,
  xe as birthYearValidator,
  We as dateValidator,
  L as dispatch,
  Ne as emailValidator,
  _e as lengthValidator,
  A as numericValidator,
  ke as phoneNumberValidator,
  Ye as postalCodeValidator,
  C as regexValidator,
  M as relay,
  F as requiredValidator
};
